Added existing item model to dedeup donations/giftcards/refunds

danthaman44 opened the pull request:
1. Created new DB migration for existing_items table to track duplicate transactions
2. Created ActiveRecord model for ExistingItem

jtmckibb commented on app/models/existing_item.rb:
```diff
@@ -0,0 +1,5 @@
+class ExistingItem < ApplicationRecord
+  validates_presence_of :idempotent_key, :item_type
```
nit, can you name it `idempotency_key`

danthaman44 commented on db/migrate/20200425033724_create_existing_items.rb:
```diff
@@ -0,0 +1,8 @@
+class CreateExistingItems < ActiveRecord::Migration[6.0]
```
Should this be named "ExistingTransaction"?

jtmckibb replied:
I'd generalize it and call it `Event`

jtmckibb replied:
because the question you're asking is: "Has this event happened already?"

jtmckibb replied:
Or maybe `ExistingEvent`

jtmckibb replied:
whatever you think — I think of Transactions as something that you make with the database, and not all calls to our API will lead to a Transaction

jtmckibb commented on db/migrate/20200425033724_create_existing_items.rb:
```diff
@@ -0,0 +1,8 @@
+  def change
+    create_table :existing_items do |t|
+      t.string :idempotent_key
+      t.integer :item_type
```
I'd just call this type instead of item_type since we already have an attribute called item_type and I wouldn't wanna confused it

jtmckibb commented on app/models/existing_item.rb:
```diff
@@ -0,0 +1,5 @@
+class ExistingItem < ApplicationRecord
+  validates_presence_of :idempotent_key, :item_type
+  validates_uniqueness_of :idempotent_key
+  enum item_type: [:donation, :giftcard, :refund]
```
`charges_create`, and `webhooks_create_refund_updated` is all we need

jtmckibb commented on app/models/existing_item.rb:
```diff
@@ -0,0 +1,5 @@
+class ExistingItem < ApplicationRecord
+  validates_presence_of :idempotent_key, :item_type
+  validates_uniqueness_of :idempotent_key
```
You should add a spec that tests that this is unique

jtmckibb replied:
and you also should add a factory

danthaman44 commented:
@jtmckibb Addressed your feedback! All tests pass:
`danthaman-mac:ruby danthaman$ heroku local:run bundle exec rspec
[OKAY] Loaded ENV .env File as KEY=VALUE Format
.......**...*.....**.........................................................***.....................................................

Pending: (Failures listed here are expected and do not affect your suite's status)

  1) CustomerMailer add some examples to (or delete) /Users/danthaman/Workspace/ruby/spec/mailers/customer_mailer_spec.rb
     # Not yet implemented
     # ./spec/mailers/customer_mailer_spec.rb:4

  2) DonationDetail add some examples to (or delete) /Users/danthaman/Workspace/ruby/spec/models/donation_detail_spec.rb
     # Not yet implemented
     # ./spec/models/donation_detail_spec.rb:4

  3) GiftCardAmount add some examples to (or delete) /Users/danthaman/Workspace/ruby/spec/models/gift_card_amount_spec.rb
     # Not yet implemented
     # ./spec/models/gift_card_amount_spec.rb:4

  4) Item add some examples to (or delete) /Users/danthaman/Workspace/ruby/spec/models/item_spec.rb
     # Not yet implemented
     # ./spec/models/item_spec.rb:4

  5) Location add some examples to (or delete) /Users/danthaman/Workspace/ruby/spec/models/location_spec.rb
     # Not yet implemented
     # ./spec/models/location_spec.rb:4

  6) Gift Cards API POST /gift_cards
     # No reason given
     # ./spec/requests/gift_cards_spec.rb:10

  7) Gift Cards API GET /gift_cards/:id
     # No reason given
     # ./spec/requests/gift_cards_spec.rb:40

  8) Gift Cards API PUT /gift_cards/:id
     # No reason given
     # ./spec/requests/gift_cards_spec.rb:66


Finished in 15.42 seconds (files took 0.90486 seconds to load)
133 examples, 0 failures, 8 pending`

jtmckibb commented:
Brilliant 👏
